satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,456 @@
+	Send(*pb.SatelliteMessage) error
+	Recv() (*pb.StorageNodeMessage, error)
+}
+
+// Endpoint for handling the transfer of pieces for Graceful Exit.
+type Endpoint struct {
+	log       *zap.Logger
+	interval  time.Duration
+	db        DB
+	overlaydb overlay.DB
+	overlay   *overlay.Service
+	metainfo  *metainfo.Service
+	orders    *orders.Service
+	config    Config
+}
+
+type pendingTransfer struct {
+	path             []byte
+	pieceSize        int64
+	satelliteMessage *pb.SatelliteMessage
+}
+
+// pendingMap for managing concurrent access to the pending transfer map.
+type pendingMap struct {
+	mu   sync.RWMutex
+	data map[storj.PieceID]*pendingTransfer
+}
+
+// newPendingMap creates a new pendingMap and instantiates the map.
+func newPendingMap() *pendingMap {
+	newData := make(map[storj.PieceID]*pendingTransfer)
+	return &pendingMap{
+		data: newData,
+	}
+}
+
+// put adds to the map.
+func (pm *pendingMap) put(pieceID storj.PieceID, pendingTransfer *pendingTransfer) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+
+	pm.data[pieceID] = pendingTransfer
+}
+
+// get returns the pending transfer item from the map, if it exists.
+func (pm *pendingMap) get(pieceID storj.PieceID) (pendingTransfer *pendingTransfer, ok bool) {
+	pm.mu.RLock()
+	defer pm.mu.RUnlock()
+
+	pendingTransfer, ok = pm.data[pieceID]
+	return pendingTransfer, ok
+}
+
+// length returns the number of elements in the map.
+func (pm *pendingMap) length() int {
+	pm.mu.RLock()
+	defer pm.mu.RUnlock()
+
+	return len(pm.data)
+}
+
+// delete removes the pending transfer item from the map.
+func (pm *pendingMap) delete(pieceID storj.PieceID) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+
+	delete(pm.data, pieceID)
+}
+
+// DRPC returns a DRPC form of the endpoint.
+func (endpoint *Endpoint) DRPC() pb.DRPCSatelliteGracefulExitServer {
+	return &drpcEndpoint{Endpoint: endpoint}
+}
+
+// NewEndpoint creates a new graceful exit endpoint.
+func NewEndpoint(log *zap.Logger, db DB, overlaydb overlay.DB, overlay *overlay.Service, metainfo *metainfo.Service, orders *orders.Service, config Config) *Endpoint {
+	return &Endpoint{
+		log:       log,
+		interval:  time.Millisecond * buildQueueMillis,
+		db:        db,
+		overlaydb: overlaydb,
+		overlay:   overlay,
+		metainfo:  metainfo,
+		orders:    orders,
+		config:    config,
+	}
+}
+
+// Process is called by storage nodes to receive pieces to transfer to new nodes and get exit status.
+func (endpoint *Endpoint) Process(stream pb.SatelliteGracefulExit_ProcessServer) error {
+	return endpoint.doProcess(stream)
+}
+
+// Process is called by storage nodes to receive pieces to transfer to new nodes and get exit status.
+func (endpoint *drpcEndpoint) Process(stream pb.DRPCSatelliteGracefulExit_ProcessStream) error {
+	return endpoint.doProcess(stream)
+}
+
+func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
+	ctx := stream.Context()
+	defer mon.Task()(&ctx)(&err)
+
+	peer, err := identity.PeerIdentityFromContext(ctx)
```

Comment:
Should we have an TODO here to handle DQed node?

Reply:
Added